src/dune_engine/build_config.mli

```diff
@@ -0,0 +1,144 @@
+(** Build system's configuration. *)
+
+open! Stdune
+open! Import
+module Action_builder = Action_builder0
+
+module Context_or_install : sig
+  type t =
+    | Install of Context_name.t
+    | Context of Context_name.t
+
+  val to_dyn : t -> Dyn.t
+end
+
+type extra_sub_directories_to_keep = Subdir_set.t
+
+module type Rule_generator = sig
+  (** The rule generator.
+
+      This callback is used to generate the rules for a given directory in the
+      corresponding build context. It receives the directory for which to
+      generate the rules and the split part of the path after the build context.
+      It must return an additional list of sub-directories to keep. This is in
+      addition to the ones that are present in the source tree and the ones that
+      already contain rules.
+
+      It is expected that [gen_rules] only generate rules whose targets are
+      descendant of [dir].
+
+      The callback should return [None] if it doesn't know about the given
+      [Context_or_install.t]. *)
+  val gen_rules :
+       Context_or_install.t
+    -> dir:Path.Build.t
+    -> string list
+    -> (extra_sub_directories_to_keep * Rules.t) option Memo.Build.t
+
+  (** [global_rules] is a way to generate rules in arbitrary directories
+      upfront. *)
+  val global_rules : Rules.t Memo.Lazy.t
+end
+
+module Error : sig
+  module Id : sig
+    type t
+
+    module Map : Map.S with type key = t
+
+    val compare : t -> t -> Ordering.t
+
+    val to_int : t -> int
+
+    val to_dyn : t -> Dyn.t
+  end
+
+  (** Errors when building a target *)
+  type t
+
+  val create : exn:Exn_with_backtrace.t -> t
+
+  val info : t -> User_message.t * User_message.t list * Path.t option
+
+  val promotion : t -> Diff_promotion.Annot.t option
+
+  val id : t -> Id.t
+end
+
+module Handler : sig
+  type event =
+    | Start
+    | Finish
+    | Fail
+    | Interrupt
+
+  type error =
+    | Add of Error.t
+    | Remove of Error.t
+
+  type t =
+    { error : error list -> unit Fiber.t
+    ; build_progress : complete:int -> remaining:int -> unit Fiber.t
+    ; build_event : event -> unit Fiber.t
+    }
+
+  val report_progress : t -> rule_done:int -> rule_total:int -> unit Fiber.t
+
+  val last_event : event option ref
+
+  val report_build_event : t -> event -> unit Fiber.t
+
+  val do_nothing : t
+
+  val create :
+       error:(error list -> unit Fiber.t)
+    -> build_progress:(complete:int -> remaining:int -> unit Fiber.t)
+    -> build_event:(event -> unit Fiber.t)
+    -> t
+end
+
+type t =
+  { contexts : Build_context.t Context_name.Map.t Memo.Lazy.t
+  ; rule_generator : (module Rule_generator)
+  ; sandboxing_preference : Sandbox_mode.t list
+  ; mutable rule_done : int
+  ; mutable rule_total : int
+  ; mutable errors : Error.t list
```

Comment:
It's not great to expose mutability in the mli. Also the name "config" seems strange to me. This record is basically the state of the build system, not some static part that was initialized.

Reply:
Yeah, I was thinking that we should perhaps split this into two parts: the immutable "config" and mutable state that will live in `build_system.ml`. Let me try to do that.

Reply:
OK, done.